src/logics/SubstLoopBreaker.cpp

```diff
@@ -0,0 +1,333 @@
+                seen.insert(var_tr, true);
+                sna[var_sr].setProcessing();
+                for (int j = 0; j < sna[var_sr].nChildren(); j++) {
+                    PTRef child_tr = sna[var_sr].getChildTerm(j);
+                    SNRef cn = SNRef_Undef;
+                    if (PTRefToSNRef.peek(child_tr, cn)) {
+                        // The child is already created
+                        if (sna[cn].getParent() == SNRef_Undef && cn != var_sr)
+                            sna[cn].setParent(var_sr);
+                    }
+                    else {
+                        // The child is not yet created, so create it.
+                        // The child cannot have a substitution target since in that case it would have been created.
+                        // No need to push it to the queue therefore.
+                        cn = sna.alloc(child_tr, PTRef_Undef);
+                    }
+                    // Update the child to var
+                    sna[var_sr][j] = cn;
+                    queue.push(cn);
+                }
+                sna[var_sr].setProcessed();
+            }
+        }
+    }
+    return minimizeRoots(std::move(roots));
+}
+
+vec<vec<SNRef>> SubstLoopBreaker::findLoops(vec<SNRef>& startNodes) {
+    vec<vec<SNRef>> loops;
+    TarjanAlgorithm tarjan(sna);
+    for (int i = 0; i < startNodes.size(); i++) {
+        vec<vec<SNRef>> loops_tmp = tarjan.getLoops(startNodes[i]);
+        for (int i = 0; i < loops_tmp.size(); i++)
+            loops.push(std::move(loops_tmp[i]));
+    }
+    return loops;
+}
+
+Map<PTRef,PtAsgn,PTRefHash> SubstLoopBreaker::constructLooplessSubstitution(const vec<Map<PTRef,PtAsgn,PTRefHash>::Pair*>&& substs)
+{
+    Map<PTRef,PtAsgn,PTRefHash> substs_out;
+    for (int i = 0; i < substs.size(); i++) {
+        auto pair = substs[i];
+        if (pair->data.sgn != l_True)
+            continue;
+
+        SNRef subst_node = sna.getSNRefBySource(pair->key);
+        if (sna[subst_node].hasChildren())
+            substs_out.insert(pair->key, pair->data);
+    }
+    return substs_out;
+}
+
+//
+// Identify and break any substitution loops
+//
+Map<PTRef,PtAsgn,PTRefHash> SubstLoopBreaker::operator() (const vec<Map<PTRef,PtAsgn,PTRefHash>::Pair*>&& substs)
+{
+    assert(seen.elems() == 0);
+
+    if (substs.size() == 0) return Map<PTRef,PtAsgn,PTRefHash>();
+
+    vec<SNRef> startNodes = constructSubstitutionGraph(std::move(substs));
+
+    while (true) {
+        vec<vec<SNRef>> loops = findLoops(startNodes);
+
+        // Terminate if no loops found
+        if (loops.size() == 0)
+            break;
+
+//        printGraphAndLoops(startNodes, loops);
+        vec<SNRef> orphans = breakLoops(loops);
+        for (int i = 0; i < orphans.size(); i++)
+            startNodes.push(orphans[i]);
+        sort(startNodes);
+        uniq(startNodes);
+//        printGraphAndLoops(startNodes, loops);
+    }
+    return constructLooplessSubstitution(std::move(substs));
```

Comment:
Applying std::move twice to the same variable is very suspicious. If you move from it once, that it is in unspecified state and should not be accessed anymore.
As I write in another comment, I believe this parameter should be just classic `const &` and the moves should be removed.

Reply:
This is fixed now.